File: csharp/AppEncryption/Crypto/CryptoPolicy.cs
using System;
using GoDaddy.Asherah.Crypto.ExtensionMethods;

namespace GoDaddy.Asherah.Crypto
{
    /// <summary>
    /// The crypto policy interface dictates the various behaviors of Asherah.
    /// </summary>
    public abstract class CryptoPolicy
    {
        public enum KeyRotationStrategy
        {
            Inline,
            Queued,
        }

        /// <summary>
        /// Checks if the key is expired.
        /// </summary>
        ///
        /// <param name="keyCreationDate">The key creation date as a <see cref="DateTimeOffset"/> object.</param>
        /// <returns><value>true</value> if the key is expired, else <value>false</value>.</returns>
        public abstract bool IsKeyExpired(DateTimeOffset keyCreationDate);

        /// <summary>
        /// Gets the key revoke check time in milliseconds.
        /// </summary>
        ///
        /// <returns>The key revoke check time in milliseconds.</returns>
        public abstract long GetRevokeCheckPeriodMillis();

        /// <summary>
        /// Checks if the <see cref="CryptoPolicy" /> allows caching of SystemKeys.
        /// </summary>
        ///
        /// <returns><value>true</value> if SystemKey caching is enabled, else <value>false</value>.</returns>
        public abstract bool CanCacheSystemKeys();

        /// <summary>
        /// Checks if the <see cref="CryptoPolicy" /> allows caching of IntermediateKeys.
        /// </summary>
        ///
        /// <returns><value>true</value> if IntermediateKey caching is enabled, else <value>false</value>.</returns>
        public abstract bool CanCacheIntermediateKeys();

        /// <summary>
        /// Checks if the <see cref="CryptoPolicy" /> allows caching of Sessions.
        /// </summary>
        ///
        /// <returns><value>true</value> if Session caching is enabled, else <value>false</value>.</returns>
        public abstract bool CanCacheSessions();

        /// <summary>
        /// Gets the maximum number of session objects that can be cached.
        /// </summary>
        ///
        /// <returns>The size of the session cache.</returns>
        public abstract long GetSessionCacheMaxSize();

        /// <summary>
        /// Gets the session cache expiry time limit in milliseconds.
        /// </summary>
        ///
        /// <returns>The session cache expiry time limit in milliseconds.</returns>
        public abstract long GetSessionCacheExpireMillis();

        /// <summary>
        /// Checks if a notification should be sent when a DRK is using an expired IK.
        /// </summary>
        ///
        /// <returns><code>true</code> if notification sending is enabled, else <code>false</code>.</returns>
        public abstract bool NotifyExpiredIntermediateKeyOnRead();

        /// <summary>
        /// Checks if a notification should be sent when an expired SK is used during read.
        /// </summary>
        ///
        /// <returns><code>true</code> if notification sending is enabled, else <code>false</code>.</returns>
        public abstract bool NotifyExpiredSystemKeyOnRead();

        /// <summary>
        /// Get the key rotation strategy.
        /// </summary>
        ///
        /// <returns>A <see cref="CryptoPolicy.KeyRotationStrategy"/> object.</returns>
        public abstract KeyRotationStrategy GetKeyRotationStrategy();

        /// <summary>
        /// Checks if the key rotation strategy is <see cref="KeyRotationStrategy.Inline"/>.
        /// </summary>
        ///
        /// <returns><code>true</code> if key rotation is <see cref="KeyRotationStrategy.Inline"/>, else <code>false</code>.</returns>
        public virtual bool IsInlineKeyRotation()
        {
            return GetKeyRotationStrategy() == KeyRotationStrategy.Inline;
        }

        /// <summary>
        /// Checks if the key rotation strategy is <see cref="KeyRotationStrategy.Queued"/>.
        /// </summary>
        ///
        /// <returns><code>true</code> if key rotation is <see cref="KeyRotationStrategy.Queued"/>, else <code>false</code>.</returns>
        public virtual bool IsQueuedKeyRotation()
        {
            return GetKeyRotationStrategy() == KeyRotationStrategy.Queued;
        }

        /// <summary>
        /// Truncate the SystemKey created time to the nearest minute.
        /// </summary>
        ///
        /// <param name="dateTimeOffset">A <see cref="DateTimeOffset"/> object.</param>
        /// <returns>A <see cref="DateTimeOffset"/> object truncated to the nearest minute. </returns>
        public virtual DateTimeOffset TruncateToSystemKeyPrecision(DateTimeOffset dateTimeOffset)
        {
            return dateTimeOffset.Truncate(TimeSpan.FromMinutes(1));
        }

        /// <summary>
        /// Truncate the IntermediateKey created time to the nearest minute.
        /// </summary>
        ///
        /// <param name="dateTimeOffset">A <see cref="DateTimeOffset"/> object.</param>
        /// <returns>A <see cref="DateTimeOffset"/> object truncated to the nearest minute. </returns>
        public virtual DateTimeOffset TruncateToIntermediateKeyPrecision(DateTimeOffset dateTimeOffset)
        {
            return dateTimeOffset.Truncate(TimeSpan.FromMinutes(1));
        }
    }
}

File: java/app-encryption/src/main/java/com/godaddy/asherah/crypto/CryptoPolicy.java
package com.godaddy.asherah.crypto;

import java.time.Instant;
import java.time.temporal.ChronoUnit;

public interface CryptoPolicy {
  enum KeyRotationStrategy {
    INLINE,
    QUEUED
  }

  /**
   * Checks if the key is expired.
   * @param keyCreationDate The key creation date as an {@link java.time.Instant} object.
   * @return {@code true} if the key is expired, else {@code false}.
   */
  boolean isKeyExpired(Instant keyCreationDate);

  /**
   * Gets the key revoke check time in milliseconds.
   * @return The key revoke check time in milliseconds.
   */
  long getRevokeCheckPeriodMillis();

  /**
   * Checks if the {@link CryptoPolicy} allows caching of SystemKeys.
   * @return {@code true} if SystemKey caching is enabled, else {@code false}.
   */
  boolean canCacheSystemKeys();

  /**
   * Checks if the {@link CryptoPolicy} allows caching of IntermediateKeys.
   * @return {@code true} if IntermediateKey caching is enabled, else {@code false}.
   */
  boolean canCacheIntermediateKeys();

  /**
   * Checks if the {@link CryptoPolicy} allows caching of {@link com.godaddy.asherah.appencryption.Session}.
   * @return {@code true} if {@link com.godaddy.asherah.appencryption.Session} caching is enabled,
   *          else {@code false}.
   */
  boolean canCacheSessions();

  /**
   * Gets the maximum number of {@link com.godaddy.asherah.appencryption.Session} objects that can be cached.
   * @return The size of the session cache.
   */
  long getSessionCacheMaxSize();

  /**
   * Gets the session cache expiry time limit in milliseconds.
   * @return The session cache expiry time limit in milliseconds.
   */
  long getSessionCacheExpireMillis();

  /**
   * Checks if a notification should be sent when a DRK is using an expired IK.
   * @return {@code true} if notification sending is enabled, else {@code false}.
   */
  boolean notifyExpiredIntermediateKeyOnRead();

  /**
   * Checks if a notification should be sent when an expired SK is used during read.
   * @return {@code true} if notification sending is enabled, else {@code false}.
   */
  boolean notifyExpiredSystemKeyOnRead();

  /**
   * Get the key rotation strategy.
   * @return A {@link CryptoPolicy.KeyRotationStrategy} object.
   */
  KeyRotationStrategy keyRotationStrategy();

  /**
   * Checks if the key rotation strategy is {@link KeyRotationStrategy#INLINE}.
   * @return {@code true} if key rotation is {@link KeyRotationStrategy#INLINE}, else {@code false}
   */
  default boolean isInlineKeyRotation() {
    return keyRotationStrategy() == KeyRotationStrategy.INLINE;
  }

  /**
   * Checks if the key rotation strategy is {@link KeyRotationStrategy#QUEUED}.
   * @return {@code true} if key rotation is {@link KeyRotationStrategy#QUEUED}, else {@code false}
   */
  default boolean isQueuedKeyRotation() {
    return keyRotationStrategy() == KeyRotationStrategy.QUEUED;
  }

  /**
   * Truncate the SystemKey created time to the nearest minute.
   * @param instant A {@link java.time.Instant} object.
   * @return A {@link java.time.Instant} object truncated to the nearest minute.
   */
  default Instant truncateToSystemKeyPrecision(Instant instant) {
    return instant.truncatedTo(ChronoUnit.MINUTES);
  }

  /**
   Truncate the IntermediateKey created time to the nearest minute.
   * @param instant A {@link java.time.Instant} object.
   * @return A {@link java.time.Instant} object truncated to the nearest minute.
   */
  default Instant truncateToIntermediateKeyPrecision(Instant instant) {
    return instant.truncatedTo(ChronoUnit.MINUTES);
  }
}
